org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/typeHierarchy/TypeHierarchyView.java

```diff
@@ -11,32 +11,343 @@
+						if (symbolsContainer != null) {
+							LSPEclipseUtils.open(symbolsContainer.file.getLocationURI().toASCIIString(), item.getSelectionRange());
+						}
+					} catch (URISyntaxException e) {
+						LanguageServerPlugin.logError(e);
+					}
+				}
+			}
+		});
+		treeViewer.addSelectionChangedListener(new ISelectionChangedListener() {
+			@Override
+			public void selectionChanged(final SelectionChangedEvent event) {
+				onHierarchySelectionChanged(event);
+			}
+		});
+
+		memberViewForm = new ViewForm(splitter, SWT.NONE);
+		Control memberControl = createMemberControl(memberViewForm);
+		memberControl.setEnabled(false);
+		memberViewForm.setContent(memberControl);
+
+		memberLabel = new CLabel(memberViewForm, SWT.NONE);
+		memberViewForm.setTopLeft(memberLabel);
+
+		infoText = new Label(pagebook, SWT.TOP | SWT.LEFT | SWT.WRAP);
+		infoText.setText(Messages.TH_diplay_hint);
+		pagebook.showPage(infoText);
+
+		FileBuffers.getTextFileBufferManager().addFileBufferListener(fileBufferListener);
+	}
+
+	@Override
+	public void dispose() {
+		FileBuffers.getTextFileBufferManager().removeFileBufferListener(fileBufferListener);
+		cachedSymbols.forEach((uri, container) -> {container.dispose();});
+		super.dispose();
+	}
+
+	public void initialize(final IDocument document, final int offset) {
+		treeViewer.setInput(new HierarchyViewInput(document, offset));
+		pagebook.showPage(splitter);
+	}
+
+	private void onHierarchySelectionChanged(SelectionChangedEvent event) {
+		var selection = event.getSelection();
+		if (selection instanceof TreeSelection && !selection.isEmpty()) {
+			var element = ((TreeSelection) selection).getFirstElement();
+			if (element instanceof TypeHierarchyItem item) {
+				typeName = item.getName();
+				IFile file = null;
+				SymbolsContainer symbolsContainer = null;
+				try {
+					symbolsContainer = cachedSymbols.get(new URI(item.getUri()));
+				} catch (URISyntaxException e) {
+					LanguageServerPlugin.logError(e);
+				}
+				if (symbolsContainer != null) {
+					file = symbolsContainer.file;
+				} else {
+					file = LSPEclipseUtils.getFileHandle(item.getUri());
+				}
+				refreshMemberViewer(file, typeName, false);
+			}
+		}
+	}
+
+	private synchronized void refreshMemberViewer(IFile file, String typeName, boolean documentModified) {
+		PlatformUI.getWorkbench().getDisplay().asyncExec(() -> {
+			if (file != null && file.exists()) {
+				refreshSymbols(getSymbolsContainer(file), documentModified);
+				var symbol = getDocumentSymbol(typeName, file);
+				memberViewer.setInput(symbol);
+				memberLabel.setText(typeName);
+				if (symbol != null ) {
+					memberViewer.getControl().setEnabled(true);
+					memberViewer.setSelection(new StructuredSelection(symbol));
+					memberLabel.setImage(symbolsLabelProvider.getImage(symbol));
+				}
+			} else {
+				memberViewer.setInput(null); // null clears it
+				var text = Messages.TH_cannot_find_file;
```

Comment:
should the text not be wrapped in a call to NLS so that it can be translated if desired? For translations to work properly, the file should be a placeholder inside the string that is the message.properties file.

Reply:
Messages already extends NLS. So I assumed that the String `TH_cannot_find_file` will be translated correctly.

Reply:
Still, one need to account for different grammar structures in different languages, so one should not make assumptions as to what those structures will be. The current code makes the assumption that the file can be just appended at the end end of the sentence, which might not be true in some languages. That is why the message should be
TH_cannot_find_file = Cannot find file {0}
so that a translator can decide where {0} will go.

Reply:
Okay, I got it. Thank you!